### ISSUE TITLE
TAF parsing

### ISSUE THREAD

**Issue by f-loris** [mivek/MetarParser#157]
At first, many thanks for this great library.

**TAF parser discrepancies**
I discovered that the `TAFParser` relies on line breaks and started to add line breaks manually for each trend since my data doesn't include any line break.

There I discovered discrepancies between NOAA and this parser, mostly related to trends and their probabilities.
Let's pick the following report from NOAA (2019/05/29 21:36):

```
TAF LSZH 292025Z 2921/3103 VRB03KT 9999 FEW020 BKN080 TX20/3014Z TN06/3003Z PROB30 
      TEMPO 2921/2923 SHRA 
      BECMG 3001/3004 4000 MIFG NSC PROB40 3003/3007 1500 BCFG SCT004 PROB30 3004/3007 0800 FG VV003 
      BECMG 3006/3009 9999 FEW030 PROB40 
      TEMPO 3012/3017 30008KT
```
For me it seems that the line breaks are not really correct in terms of reflecting each trend. For example in line 3 there is a `BECMG` trend  and 2 `PROB` trends. In line 4, the ending `PROB40` already accounts to the `TEMPO` in line 5. Therefore, I see it as dangerous to rely on line breaks in the report.

My interpretation of this TAF report is based on the parsed version provided by the NOAA. This can be found attached [taf_example.zip](https://github.com/mivek/MetarParser/files/3236210/taf_example.zip)
I also checked this document provided by the EASA: https://www.easa.europa.eu/sites/default/files/dfu/NPA%202014-07%20%28A%29.pdf

**Describe the solution you'd like**
Based on my interpretation, I'd suggest the following changes:

1. Don't rely on line breaks for parsing. Instead, extract trends by looking for `TEMPO | BECMG | FM | PROB` in the report.
2. Based on my findings `TEMPO` can have a `PROB` as prefix, therefore something like `PROB40 
      TEMPO` should result in a `TEMPO` trend with a probability set to 40. So, this would also be an additional case for point 1 (e.g. the following regex `PROB\d{2}\sTEMPO|TEMPO|BECMG|FM|PROB`) 
3. Based on my understanding `PROB` can't be in the middle of any trend, so checking the content further for it is not needed. Therefore, I think the following checks are not needed/correct:
https://github.com/mivek/MetarParser/blob/266ecb58b1c4343e52442142f6d61f76e0008060/src/main/java/io/github/mivek/parser/TAFParser.java#L96
https://github.com/mivek/MetarParser/blob/266ecb58b1c4343e52442142f6d61f76e0008060/src/main/java/io/github/mivek/parser/TAFParser.java#L151
Also, only PROB and TEMPO (optional) can have a probability. The TAF itself and the other trends not.

I'd really appreciate feedback on that topic, since I'm also no real expert on TAF reports. I'm also willing to contribute on this topic.

**Comment by mivek** [comment]
@f-loris thank you for your feed back.

Indeed, the parsing of TAF messages relies on line breaks mostly because of the NOAA format.
I think it would be possible not to rely on the line breaks to parse TAF but it implies a refactor of the parse method of the TAFParser. Instead of parsing each line as a trend, the parser should parse tokens and create a new trend when the token is `BECMG` or `FM` or `TEMPO`.

The important point is to insure the compatibility with the NOAA format but that would be in the TAFFacade I think.

Regarding the probability of trends I added the attribute  on all trends because I did not found documentation on the topic. I will read the document you provided and search on that topic too.

**Comment by f-loris** [comment]
@mivek Right, it should be tokenized based on the trend, but not only `BECMG` or `FM` or `TEMPO`. In addition to them, it should be also tokenized by `PROB\d{2}` or `PROB\d{2}\sTEMPO`. 

Yes, it should definitely be compatible with the NOAA format. Actually, when you compare the decoded form from NOAA (the one I attached), it currently produces a different result. Therefore, when the points I mentioned are implemented, it should produce the same results. 

Should I file a PR? Should I wait until https://github.com/mivek/MetarParser/pull/156/ is finished?

**Comment by mivek** [comment]
@f-loris I merged #156 into master.

I think we should add a new abstract class (AbstractTafProbTrend? I am not sure about the naming yet) extending `AbstractTafTrend` and make `PROBTafTrend` and `TEMPOTafTrend` extending AbstractTafProbTrend, move the probability attribute to this class.

Then the methods `parse` and `processLines` will have to be updated.

You can file a PR whenever you want, I won't be available to work on this topic before mid-June.
Thank you for your help

**Comment by f-loris** [comment]
Great thx!

Sounds good. I plan to start with this beginning of next week and file a PR.  Then we can continue discussion.
